Bugfix/subsref

Nabarb opened the pull request:
I've rewritten the subsref of tank and animal to take care of some unexpected behaviour. A Block revision will follow soon (tomorrow).

To summarize:
 the {} operator can be used to access the children object in the layer below or two layers below. That means that you can access blocks and animals directly from the tank level.
## Classic usage
`tankObj{aa}` is equivalent to` tankObj.Children(aa)`
`tankObj{aa,bb} `is equivalent to` tankObj.Children(aa).Children(bb)`

**animalObj changes it's behaviour** depending on the fact that's a scalar or a vector:
- Scalar case
`animalObj{bb}` is equivalent to` animalObj.Children(bb)`;

`animalObj(bb,cc)` is unsupported. Issues a warning and ignores all the indexes except bb, that means that is equivalent to `animalObj.Children(bb)`.

- Vector case
`animalObj(bb)` is unsupported. Issues an error. (with one exception, more on this later)*;

`animalObj(aa,bb)` is equivalent to `animalObj(aa).Children(bb).`
This is equivalent to `tankObj{aa,bb}`

## Advanced Indexing:
- You can address **multiple objects** at the same time, just use vectors as indexes.

`aa = [1,2]; bb = 1; tankObj{aa,bb}` 
will return the first block from the first two animals.

`aa = [1,2]; bb = [1,3]; tankObj{aa,bb} `
will return the first and third block from the first two animals.

- Indexes ('aa' and 'bb') can be numbers or Public Keys.
They work interchangeably: this means you can use numerical indexes as first index and  keys as second indexes. **To address multiple objects at once using keys, stack them in cell arrays.**

`aa=[1,2]; bb=key1; tankObj{aa,bb} `
will look for key1 in all the blocks in animal 1 and 2.

`aa = [1,2]; bb={key1,key2}; tankObj{aa,bb}`
 will look for key1 and key2 in all the blocks in animal 1 and 2.

- Asymmetrical indexing
You can access different indexes in different animals using curly brackets.
`tankObj{[1,2],{1,[1,2]}}` gives you access from the first block from the first animal, and the first two from the second one, 

You can also use keys.
`tankObj{[1,2], {key1, {key2 key3} } }` looks for key1 between the children of animal1 and for key2 and key3 between the children of animal2. 

`tankObj{ {keya1, keya2} , {keyb1, {keyb2 keyb3} } }` is _technically_ supported but needs more testing and may lead to unexpected results.

`tankObj{ [1,2], {1, {key1 key2} } }` is **not** supported yet.

Nabarb committed:
Issue #67 solved

This fixes both the infinite loop and the error on direct pars loading.
Corrected a typo and swapped a line, nothing to see here!

Nabarb committed:
Basically rewritten the subsref of tank and animal

I've rewritten the subsref of tank and animal to take care of some unexpected behavior. A Block revision will follow soon.

To summarize:
 the {} operator can be used to access the children object in the layer below or two layers below. That means that you can access blocks and animals directly from the tank level.
eg
tankObj{aa} is equivalent to tankObj.Children(aa)
tankObj{aa,bb} is equivalent to tankObj.Children(aa).Children(bb)

animalObj changes it's behavior depending on the fact that's a scalar or a vector:
Scalar case
animalObj{bb} is equivalent to animalObj.Children(bb)
animalObj(bb,cc) is unsupported. Issues a warning and ignores all the indexes except bb, that means that is equivalent to animalObj.Children(bb).

Vector case
animalObj(bb) is unsupported. Issues an error.*
animalObj(aa,bb) is equivalent to animalObj(aa).Children(bb).
This is equivalent to tankObj{aa,bb}

Advanced Indexing:
You can address multiple objects at the same time, just use vectors as indexes.
E.g. aa = [1,2]; bb = 1; tankObj{aa,bb} will return the first block from the first two animals.
aa = [1,2]; bb = [1,3]; tankObj{aa,bb} will return the first and third block from the first two animals.

Indexes ('aa' and 'bb') can be numbers or Public Keys.
They work both interchangeably: this means you can use numerical indexes as first index and  keys as second index. To address multiple objects at once using keys, stack them in cell arrays.
E.g. aa=[1,2]; bb=key1; tankObj{aa,bb} will look for key1 in all the blocks in animal 1 and 2.
aa = [1,2]; bb={key1,key2}; tankObj{aa,bb} will look for key1 and key2 in all the blocks in animal 1 and 2.

Asymmetrical indexing. This will be documented later on.

Nabarb committed:
defaults to test demo_nigelab + small bugfix

Updated defaults to test everything properly.

Swapped the order of a if elseif statemetn for advanced indexing.

Nabarb committed:
blockObj subsref update

BlockObj subsref is now working as intended (I believe)

By using {} in conjunction with defaults.Shortcuts, it is possible to access elements two levels under the current hierarchy level.

- Shortcurts now has a new format.
The struct format uses the shortcut keyword as the struct fieldname, Below that you can customize the fields to access and specify if those fields are indexable or not.

E.g let's say you wanted to access the first 100 samples in the Raw data. You would have to write:
blockObj.Channels(iCh).Raw(1:100);
With the shortcuts you configure the defaults by adding :
pars.raw.subfields = {'Channels', 'Raw'};
 pars.raw.indexable = [true      , true];

to the config file. Once that's done you can just call
blockObj{'raw',iCh,1:100};

let's explore another case. Let's say you wanted to access
blockObj.Streams.DigIO(iCh).data(1:100);

You would need to add to the config file
   pars.digIO.subsfield = {'Streams', 'DigIO', 'data'};
   pars.digIO.indexable = [false    , true   , true];
Streams is a not indexable field because is not vectorial. No () follows it.
Now typing
 blockObj{'digIO',iCh,1:100};
will lead to the desired result.

- Advanced indexing.
Doc will be added later.

- Direct Channels access
All fields in Channels are also accessible directly from block:
E.g. blockObj.raw(iCh,1:100) is equivalent to blockObj.Channels(iCh).Raw(1:100);

The operator **end** is not available in shortcuts!

Nabarb commented:
BlockObj subsref is now working as intended (I believe)

By using {} in conjunction with defaults.Shortcuts, it is possible to access directly from block elements two levels below the block hierarchy level. 

- `default.Shortcurts` now has a new format.
`default.Shortcuts` now return a struct that uses the shortcut keyword as the struct fieldname, Below that you can customize which fields to access and specify if those fields are indexable or not. See the example below.

E.g. let's say you wanted to access the first 100 samples in the Raw data of the channel iCh. You would have to write: 
`blockObj.Channels(iCh).Raw(1:100);`
With the shortcuts you configure the defaults by adding: 
`pars.raw.subfields = {'Channels', 'Raw'};`
`pars.raw.indexable = [true      , true];`
to the config file. Once that's done you can just call
`blockObj{'raw',iCh,1:100};`
You can see how the shortcut keyword is the same as the corresponding fieldname in the shortcuts struct.

Let's explore another case. Let's say you wanted to access
`blockObj.Streams.DigIO(iCh).data(1:100);`
You would need to add to the config file
  `pars.digIO.subsfield = {'Streams', 'DigIO', 'data'};`
   `pars.digIO.indexable = [false    , true   , true];`
Streams is a not an indexable field because is not vectorial. No () follows it.
Now by typing
 `blockObj{'digIO',iCh,1:100};`
You will reach the same result.

- Direct Channels access
All fields in Channels defined in Fields in the block defualt file are now also accessible directly from block:
E.g. `blockObj.raw(iCh,1:100)` is equivalent to `blockObj.Channels(iCh).Raw(1:100);`

### Advanced indexing.
**It is only available in {}, not in direct access!**
- Symmetrical multiple indexing
`blockObj{'raw',[1,2], 1:10}` will return the first ten samples from the first two channels.
- Asymmetrical indexing
You can access different indexes in different channels using curly brackets.
`blockObj{'raw',[1,2],{1,[1,2]}}` gives you access from the first sample from the first channel, and the first two from the second one.

### Warning 
The operator **end** is not available in shortcuts!

m053m716 commented on +nigeLab/+defaults/Animal.m:
```diff
@@ -22,18 +22,18 @@
-pars.NamingConvention={'SurgYear','SurgNumber'};
+pars.DynamicVarExp={'$AnimalID'};
+pars.NamingConvention={'AnimalID'};
 
 pars.SpecialMeta = struct;
-pars.SpecialMeta.SpecialVars = {'AnimalID'};
+pars.SpecialMeta.SpecialVars = {};
 pars.SpecialMeta.AnimalID.cat = '-'; % Concatenater (if used) for names
 pars.SpecialMeta.AnimalID.vars = {'SurgYear','SurgNumber'}; % KUMC "standard"
 % pars.SpecialMeta.AnimalID.vars = {'Project','SurgNumber'}; % KUMC "RC"
 
 pars.Delimiter   = '-'; % delimiter for variables in ANIMAL name
 pars.Concatenater = '-'; % concatenater for variables INCLUDED in ANIMAL name
-pars.VarExprDelimiter = {'-','_'}; % Delimiter for parsing "special" vars
+pars.VarExprDelimiter = {'_'}; % Delimiter for parsing "special" vars
```
@Nabarb have you tried leaving it this way? Or does it have to be different for MultiAnimals to work? We should address that if it is the case.

Nabarb replied:
Well, my naming scheme is different and doesn't use the surgYear-SurgNumber thing  ¯\_(ツ)_/¯.
Sorry I always forget not to push updates to the config files... Maybe we might want to make local copies of them at some point.

m053m716 replied:
Not a problem! I forgot you do the K## or whatever instead.

m053m716 commented on +nigeLab/+defaults/SD.m:
```diff
@@ -41,7 +41,7 @@
 pars.MULTCOEFF       = 4.5;  % Multiplication coefficient for noise
 pars.PKDURATION      = 1.0;  % Pulse lifetime period (suggest: 2 ms MAX)
-pars.REFRTIME        = 2.0;  % Refractory period (suggest: 2 ms MAX).
+pars.REFRTIME        = 0.5;  % Refractory period (suggest: 2 ms MAX).
```
@Nabarb Was this parameter causing issues? I think I had empirically found 2ms to work pretty well for the behavior data I'd looked at, but it probably also depends on what is called a spike. Mostly curious on your thoughts here.

Nabarb replied:
Sorry, my bad. I was testing stuff, 2 is good!

m053m716 commented on +nigeLab/+defaults/Shortcuts.m:
```diff
@@ -18,13 +18,29 @@
-   pars.spk = 'Channels(%d).Spikes';
-   pars.srt = 'Channels(%d).Sorted';
-   pars.clst = 'Channels(%d).Clusters';
+   pars.raw.subfields = {'Channels', 'Raw'};
+   pars.raw.indexable = [true      , true];
+   
+   pars.filt.subfields = {'Channels', 'Filt'};
+   pars.filt.indexable = [true      , true];
+   
+   pars.car.subfields = {'Channels', 'CAR'};
+   pars.car.indexable = [true     , true];
+   
+   pars.lfp.subfields = {'Channels', 'LFP'};
+   pars.lfp.indexable = [true     , true];
+   
+   pars.spk.subfields = {'Channels', 'Spikes'};
+   pars.spk.indexable = [true      , true];
+   
+   pars.srt.subfields = {'Channels', 'Sorted'};
+   pars.srt.indexable = [true      , true];
+   
+   pars.clst.subfields = {'Channels', 'Clusters'};
+   pars.clst.indexable = [true      , true];
+   
+   pars.digIO.subsfield = {'Streams', 'DigIO', 'data'};
+   pars.digIO.indexable = [false    , true   , true];
```
I think this is a very clear way to go about it. Nicely done.

m053m716 commented on +nigeLab/@Animal/Animal.m:
```diff
@@ -93,6 +93,7 @@
             error(['nigeLab:' mfilename ':initFailed'],...
                'Could not initialize ANIMAL object.');
          end
+         animalObj.Key = nigeLab.nigelObj.InitKey;
```
@Nabarb we had discussed a "hashing" function for relating objects to their "superiors." One thought that I had was it could be easily done based on a combination of Indexing and the field or property name. It is not relevant to the review of this branch, but bringing it to your attention here since this line is highlighted, and I was thinking about it earlier today.

For example, you could do:
```(Matlab)
% parentKey = .... (some key passed from "Parent" object via method or on constructor)
hashedID = 
animalObj.Key = horzcat(parentKey, hashedID);
```
For example, I think in `nigeLab.utils.initChannelStruct` I have added most of those structs to have their own `Key` field. In that case, we can for example do:
```(Matlab)
% In this example, want to set a unique key for an Events element
pubKey = getKey(blockObj,'Public');
keyDelim = '::'; % Or from blockObj.Pars.Experiment or whatever.
F = fieldnames(blockObj.Events);
for iF = 1:numel(F)
   field_ = F{iF};
   for iSub = 1:numel(blockObj.Events.(field_)
     % Encode the field name and index into the key:
     hashedID = char(double(field_) + iSub);
     blockObj.Events.(field_)(iSub).Key = horzcat(pubKey,keyDelim,hashedID);
   end
end
```

What do you think?

m053m716 commented on +nigeLab/@Block/Block.m:
```diff
@@ -167,10 +167,22 @@
       end
 
+      function ind = end(obj,k,n)
+          if n>2
+              % called by {}
+              warning(['nigeLab:' mfilename 'unsupportedOperator'],...
+                  '""end"" operator not supported in this context.\nThis may lead to unexpexted behavior.');
+              ind = builtin('end',obj,k,n);
+          else
+              ind = builtin('end',obj,k,n);
+          end
+      end
+      
       % Overloaded NUMARGUMENTSFROMSUBSCRIPT method for parsing indexing.
-      function n = numArgumentsFromSubscript(blockObj,s,indexingContext)
+      function n = numArgumentsFromSubscriptUNUSED(blockObj,s,indexingContext)
          % NUMARGUMENTSFROMSUBSCRIPT  Parse # args based on subscript type
          %
          %  n = blockObj.numArgumentsFromSubscript(s,indexingContext);
```
I'd be fine with just deleting it.

m053m716 commented on +nigeLab/@Block/Block.m:
```diff
@@ -167,10 +167,22 @@
 
+      function ind = end(obj,k,n)
+          if n>2
+              % called by {}
```
Yay! Thank you!

m053m716 commented on +nigeLab/@Tank/subsref.m:
```diff
@@ -58,22 +58,54 @@
             '''.'' or ''()'' references.']);
```
Something weird happens with `subsref` if I do `addChild` and add a copy of a Block that is on a different parent to the wrong Animal. Then I get errors on the `demo` data using commands you put in the `docs`:
```(matlab)
>> b = tankObj{1,1}; % As expected
>> key = getKey(b,'Public'); % As expected
>> B = tankObj{[1,2],key}; % Sorry I broke it
Index exceeds matrix dimensions.

Error in nigeLab.Animal/subsref>@(x,idx)subsref(x.Children,s(idx,:)) (line 203)
               out = arrayfun(@(x,idx) subsref(x.Children,s(idx,:)),an(:),(1:numel(an))','UniformOutput',false);

Error in nigeLab.Animal/subsref (line 203)
               out = arrayfun(@(x,idx) subsref(x.Children,s(idx,:)),an(:),(1:numel(an))','UniformOutput',false);

Error in nigeLab.Tank/subsref (line 109)
            out = subsref(tankObj.Children,s);
```

m053m716 replied:
Even without the `addChild` bit, I am still getting the following error:
```(matlab)
>> key = getKey(b,'Public');
>> B = tankObj{[1,2],key}
Index exceeds matrix dimensions.

Error in nigeLab.Animal/subsref>@(x,idx)subsref(x.Children,s(idx,:)) (line 203)
               out = arrayfun(@(x,idx) subsref(x.Children,s(idx,:)),an(:),(1:numel(an))','UniformOutput',false);

Error in nigeLab.Animal/subsref (line 203)
               out = arrayfun(@(x,idx) subsref(x.Children,s(idx,:)),an(:),(1:numel(an))','UniformOutput',false);

Error in nigeLab.Tank/subsref (line 109)
            out = subsref(tankObj.Children,s);
```

Nabarb replied:
Fixed! Nice catch bug catcher.

m053m716 commented on +nigeLab/@Animal/subsref.m:
```diff
@@ -39,186 +39,294 @@
+            % % %          end
+            % % %
+            % If more than one animalObj in array
+        else
```
I would just double-check something you put on the `docs`. I've ported a lot of your comments to the `wiki` (**[here](https://github.com/m053m716/ePhys_packages/wiki/Startup_Shortcuts#vector-case)**). Specifically, take a look at that section; at least when I reference it using the `()` indexing it does still work, I am just confused by the `docs` I guess.

m053m716 replied:
I've updated the `wiki` to what I think it should be.

m053m716 commented on +nigeLab/@Block/subsref.m:
```diff
@@ -99,108 +97,127 @@
+                   'Dimensions of arrays being concatenated are not consistent.');
+          end
+          
+          switch checkSamplesInput(subs{jj})
```
I like this helper function to keep it simpler.

m053m716 commented on +nigeLab/@Block/subsref.m:
```diff
@@ -99,108 +97,127 @@
+         
+      end
+      index = ones(1,length(TargetObj)).*(1:size(s,1));
+      varargout = {arrayfun(@(x,idx) subsref(x,s(idx,:)),TargetObj,index,'UniformOutput',false)};
```
One thing that's somewhat minor but I could see it being problematic in the future: by convention we have been trying to use row arrays for everything since Matlab `for` loops iterate on individual elements (columns) of a row vector. So we should be consistent with how the vectors are returned: currently `subsref` is returning column vectors--can we switch it so it automatically returns `row` vectors?

m053m716 replied:
This has been addressed with most recent `Push`

m053m716 reviewed (changes requested):
Overall seems really convenient. I added a lot to the **[`wiki`](https://github.com/m053m716/ePhys_packages/wiki/Startup_Shortcuts)** and did some formatting and clarification of examples as well. There are some bugs, and I have also made some changes on my local branch although I am not completely finished testing. But I need to leave for the evening so wanted to get this up before I go.

m053m716 committed:
Bugfixes, updates

------------------
# Bugfixes #
** Bugfixes **
## Major ##
** Major Bugs ** 
* Fixed issue where `B = tankObj{[1,2],key};` was throwing an error due to improper arrayfun indexing of substruct argument array.

## Minor ##
** Minor Bugs **
* `nigeLab.nigelObj/getPropertyGroups`
  + Call of tankObj{1} threw error because of unexpected naming convention used on demo dataset. Changed method to account for case where '+', '&', or '|' is used in addition to ' ' or '-' characters.

* `nigeLab.nigelObj/addPropListeners`
  + Getting an error I hadn't seen previously where it couldn't find the `checkBlocksForClones` or `checkAnimalsForClones` methods. Changed syntax in listener call and that fixed it even though effectively it should be the same reference? Not sure if this is related to subsref changes yet.

------------------
# Updates #
** Updates **
## Notable ##
** Notable Updates **
* Moved `IsRightKeyFormat` to be a method of `nigeLab.nigelObj`.

## Minor ##
** Minor Updates **
* Changed formatting on some comments, as well as indent stuff that I'm apparently retentive about.
* Fixed typos

m053m716 commented:
I've pushed some changes. Overall seems like things are working as intended. I found one bug that was potentially problematic but that's been resolved.

------------------
# Bugfixes #
## Major ##
* Fixed issue where `B = tankObj{[1,2],key};` was throwing an error due to improper arrayfun indexing of substruct argument array.

## Minor ##
* `nigeLab.nigelObj/getPropertyGroups`
  + Call of tankObj{1} threw error because of unexpected naming convention used on demo dataset. Changed method to account for case where '+', '&', or '|' is used in addition to ' ' or '-' characters.

* `nigeLab.nigelObj/addPropListeners`
  + Getting an error I hadn't seen previously where it couldn't find the `checkBlocksForClones` or `checkAnimalsForClones` methods. Changed syntax in listener call and that fixed it even though effectively it should be the same reference? Not sure if this is related to subsref changes yet.

------------------
# Updates #
## Notable ##
* Moved `IsRightKeyFormat` to be a method of `nigeLab.nigelObj`.

## Minor ##
* Changed formatting on some comments, as well as indent stuff that I'm apparently retentive about.
* Fixed typos
